### PR TITLE
Constant fold string comparisons

### PR DESCRIPTION
**Pull request by zoecarver**
This patch adds constant folding for string comparisons. Most if not all string comparisons that compare constant values can now happen at compile time. This works for both switch and if statements. 

### Input
```
func test() -> Int {
  let i = "3"
  switch i {
      case "1": return 1
      case "2": return 2
      default: return 3
  }
}
```
### Old output
```
	.cfi_startproc
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset %rbp, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register %rbp
	pushq	%rbx
	pushq	%rax
	.cfi_offset %rbx, -24
	movabsq	$-2233785415175766016, %rbx
	movl	$51, %edi
	movq	%rbx, %rsi
	callq	_$ss12_SmallStringV15_invariantCheckyyF
	movl	$51, %edi
	movq	%rbx, %rsi
	callq	_$ss13_StringObjectV15_invariantCheckyyF
	movl	$51, %edi
	movq	%rbx, %rsi
	callq	_$ss11_StringGutsV15_invariantCheckyyF
	movl	$51, %edi
	movq	%rbx, %rsi
	callq	_$sSS15_invariantCheckyyF
	movl	$50, %edi
	movq	%rbx, %rsi
	callq	_$ss12_SmallStringV15_invariantCheckyyF
	movl	$50, %edi
	movq	%rbx, %rsi
	callq	_$ss13_StringObjectV15_invariantCheckyyF
	movl	$50, %edi
	movq	%rbx, %rsi
	callq	_$ss11_StringGutsV15_invariantCheckyyF
	movl	$50, %edi
	movq	%rbx, %rsi
	callq	_$sSS15_invariantCheckyyF
	movl	$50, %edi
	movl	$51, %edx
	movq	%rbx, %rsi
	movq	%rbx, %rcx
	xorl	%r8d, %r8d
	callq	_$ss27_stringCompareWithSmolCheck__9expectingSbs11_StringGutsV_ADs01_G16ComparisonResultOtF
	movzbl	%al, %eax
	andl	$1, %eax
	xorq	$3, %rax
	addq	$8, %rsp
	popq	%rbx
	popq	%rbp
	retq
	.cfi_endproc
```
### New output
```
	pushq	%rbp
	movq	%rsp, %rbp
	movl	$3, %eax
	popq	%rbp
	retq
```

## TODO
- [ ] Where is the correct place for this function to go?
- [ ] How do I correctly query for the string comparison function?
- [ ] Tests? Where should my tests go and are there any examples so I can figure out how to write them?

### CODE REVIEW

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 1]
```diff
@@ -292,6 +293,44 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
...
+        // The string is created using a function that is passed a string literal
+        if (auto *MakeStr = dyn_cast<ApplyInst>(Defined->get())) {
+          // Get the string literal which is the first argument
+          if (auto *SL = dyn_cast<StringLiteralInst>(MakeStr->getArgumentOperands()[0].get())) {
+            if (FirstArg.empty()) {
+              FirstArg = SL->getValue();
+              continue;
+            }
+            
+            APInt IsSame(1, FirstArg == SL->getValue());
```
Comment (jckarter) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: cc @milseman. This is cool! Unfortunately, constant folding string equality is not that cut and dry because it takes Unicode canonical equivalence into account, but this is only going to check that the strings contain the same code units. Without bringing ICU entirely into the compiler itself, we could constant fold cases where the literal is identical as definitely true, and true or false cases where the string is ASCII so we know there are no canonical equivalence issues. Otherwise, we should leave the call as is.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 2]
```diff
@@ -292,6 +293,44 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
                    "the switch_enum.");
 }
 
+void BBEnumTagDataflowState::handleStringCmp(SILBasicBlock *Pred) {
+  // Find all apply instructions
+  for (auto begin = Pred->begin(); begin != Pred->end(); ++begin) {
+    if (auto *AI = dyn_cast<ApplyInst>(begin)) {
+      if (auto *FN = dyn_cast<FunctionRefInst>(AI->getCalleeOrigin())) {
+        // Only keep going if this is an apply instruction
+        if (FN->getReferencedFunctionOrNull()->getName() != "$sSS2eeoiySbSS_SStFZ") continue;
```
Comment (jckarter) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Our usual way of flagging functions for SIL optimization is to use the `@_semantics` attribute to put a unique string tag on the declaration, and to have the optimizer look for that tag instead of for a specific symbol name. `@_semantics` will delay inlining the function so that higher-level optimizations have a change to apply first, and makes the pass less sensitive to changes in the standard library itself that might affect mangled names.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 1 above)
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Thanks for explaining the issue, that makes sense. I'll add a check to ensure we are looking at an ASCII string literal.

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: This should now be resolved :)

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Do you know what function I should apply the `@_semantics` decorator to? Is [`_stringCompare`](https://github.com/apple/swift/blob/e9d4687e31a3ae8e90604d3b15bf8b241479c211/stdlib/public/core/StringComparison.swift#L17) always the function used to compare strings?

Comment (jckarter) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @milseman would know best.

**Comment by gottesmm** [comment]
I would put this into constant propagantion or constexpr.

**Comment by zoecarver** [comment]
@gottesmm I agree, `ConstExpr.cpp` seems like the right place for this. Where should I "hook in" the new function? I tried putting it in the `run` method in `ConstantEvaluatorTester.cpp` but, it never seemed to get invoked.

**Comment by gottesmm** [comment]
@zoecarver I am not sure if constexpr is ready TBH. You may need to put it in SILCombine or mandatory combiner or something like that. I would talk to @ravikandhadai

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 2 above)
Comment (milseman) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: ` _stringCompare` works, though it's already inlinable for the same-struct-bits fast-path.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 3]
```diff
@@ -292,6 +293,50 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
...
+      } else continue;
+      
+      StringRef FirstArg; // Keep track of one of the arguments
+      
+      for (auto& Arg : AI->getArgumentOperands()) {
+        ValueBase *V = (ValueBase*)Arg.get().getOpaqueValue();
+        Operand * Defined = *V->use_begin(); // Find where the argument was defined
+
+        // The string is created using a function that is passed a string literal
+        if (auto *MakeStr = dyn_cast<ApplyInst>(Defined->get())) {
```
Comment (rjmccall) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: This is all just `if (auto *MakeStr = dyn_cast<ApplyInst>(Arg))`.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 4]
```diff
@@ -292,6 +293,50 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
...
+  for (auto begin = Pred->begin(); begin != Pred->end(); ++begin) {
+    if (auto *AI = dyn_cast<ApplyInst>(begin)) {
+      if (auto *FN = dyn_cast<FunctionRefInst>(AI->getCalleeOrigin())) {
+        // Only keep going if this is an apply instruction
+        if (FN->getReferencedFunctionOrNull()->getName() != "$sSS2eeoiySbSS_SStFZ") continue;
+      } else continue;
+      
+      StringRef FirstArg; // Keep track of one of the arguments
+      
+      for (auto& Arg : AI->getArgumentOperands()) {
```
Comment (rjmccall) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: 1. Please create a helper function somewhere to check if a `SILValue` is a call to construct a constant string and, if so, to return the constructed string.
2. Please make that call properly look for calls to a specific function (or a function with the right `@_semantics` rather than assuming that all calls which take `StringLiteralInsts` as their first argument fit your conditions.
3. You can then just call that helper function on the two arguments to the comparison function instead of doing this loop.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 5]
```diff
@@ -292,6 +293,50 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
...
+              if (IsAscii->getValue() != 1) continue;
+            } else continue;
+
+            APInt IsSame(1, FirstArg == SL->getValue());
+            SILBuilder B(AI);
+            SILType IntBoolTy = SILType::getBuiltinIntegerType(1, B.getASTContext());
+            auto C1 = B.createIntegerLiteral(AI->getLoc(), IntBoolTy, IsSame);
+            auto TrueStruct = B.createStruct(AI->getLoc(), AI->getType(), {C1});
+            
+            AI->replaceAllUsesWith(TrueStruct);
```
Comment (rjmccall) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @gottesmm Should `AI` be immediately flagged for deletion here, or is it okay to leave that to happen implicitly?

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 6]
```diff
@@ -292,6 +293,50 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
                    "the switch_enum.");
 }
 
+void BBEnumTagDataflowState::handleStringCmp(SILBasicBlock *Pred) {
```
Comment (rjmccall) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Please spell out this function name.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 4 above)
Comment (ravikandhadai) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: I was just about to mention the same thing as @rjmccall. You can use this helper function: https://github.com/apple/swift/blob/91669cec8a7d34fa603c836b64bfd679abb8f1ad/lib/SILOptimizer/Mandatory/OSLogOptimization.cpp#L111 to check if this is a call to a string initializer that takes a string literal. This function can be factored out and put in a common place. I think `swift/lib/SILOptimizer/Utils/ConstExpr.h/cpp` can have this utility function.

**Comment by jckarter** [comment]
@swift-ci Please benchmark

**Comment by gottesmm** [comment]
@rjmccall I would wait on going through the nitty gritty details. I think that @zoecarver is moving this out of SILCodeMotion into SILCombine or something like that... (or am I confused)?

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 4 above)
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @ravikandhadai and @rjmccall let me know what you think of `StringLiteralInfo`.

**Comment by zoecarver** [comment]
Are those performance improvements significant? I don't have a reference for what they should look like.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Will this optimization sometimes be missed if `_stringCompare` is inlined, then?

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 1 above)
Comment (ravikandhadai) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: > Unfortunately, constant folding string equality is not that cut and dry. Equality takes Unicode canonical equivalence into account, but your check here is only going to check that the strings contain the same code units, so it could incorrectly end up turning things like "e\u{301}" == "é" into false. Without bringing ICU entirely into the compiler itself, we could limit this so that it only constant folds cases where the literal is identical as definitely true, or ASCII literals where we know there are no canonical equivalence issues. Otherwise, we should leave the call as is.

@jckarter @milseman  I have a follow up question about this comparison of unicode strings. I had added support for `string.equals` in the constant evaluator sometime back, which uses `std::string ==` to equate two string literals encountered while evaluating SIL  (which are UTF8 string as I guess we are only using UTF8 for literals in Swift.) It seems `std::string ==` does a byte by byte comparison. Is it correct to do this and fold string literals even in the UTF8 case? It does handle your example `"e\u{301}" == "é"` correctly. It also appears that `_stringCompare` does a  `memcmp` to compare strings that are `fastUTF8`. 

In other words, assuming that all `string_literal ` strings in the  SIL are UTF8, would it suffice to do a raw byte comparison?

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 2 above)
Comment (ravikandhadai) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @zoecarver What you say is possible. On the upside, inlining may also result in the code getting optimized through other optimizations. But, on a different note, it is probably more important to *not* to compare the mangled name of `_stringCompare` in the code, but add an @_semantics attribute to that function and use that instead (as suggested by others earlier). 

But, one thing to note is that, adding a @_semantics attribute to a function will interfere with the inlining of the function, which could make uses of this function in other contexts slower. So, I would think it is best to create a new PR that just adds  @_semantics attribute, and run benchmark test on it to see if there are any performance regressions. Once that lands, you can use it to modify this code here.

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: It looks like the string compares I had been optimizing were not actually using `_stringCompare` and instead used another function that took the string literals directly as arguments. `_stringCompare` takes `_StringGuts` types which require store/load. That makes it much more difficult to extract the string literal. Any ideas as to the function it is calling instead? Is it OK to rely on it always optimizing into this function?

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Turns out the function it calls is just the `op==`. Looking for `hasSemanticsAttr("string.equals")` works for both if and switch statements. Using that fixes the concern about inlining `_stringCompare` too. Any reasons not to just look for `string.equals`?

**Comment by jckarter** [comment]
I was just curious. I wasn't expecting any specific changes from this patch.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 1 above)
Comment (rjmccall) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Yeah, equality tests will have this same problem, so I guess it's a good thing that that the constant evaluator is currently disabled.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 2 above)
Comment (jckarter) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: If the semantics tag already exists, it seems reasonable to use it here.

**Comment by zoecarver** [comment]
Ping, how does this look? I am going to make a similar patch for string concatenation but, I want to get this reviewed first.

**Comment by gottesmm** [comment]
We already support String Concatenation in the constant propagation pass.

**Comment by zoecarver** [comment]
Aha, you're right!

**Comment by milseman** [comment]
@zoecarver can you add optimization tests to ensure the switches get constant folded (and that this doesn't regress)? See test/SILOptimizer/string_literals.swift and test/SILOptimizer/string_switch.swift for inspiration.

Otherwise, I'll try to find someone to assign this to.

**Comment by gottesmm** [comment]
Hey @zoecarver. I am happy to help you get this thing in! I am currently finishing up an LLVM talk that I am going to give next week. After that, we can get this ball rolling/etc.

**Comment by ravikandhadai** [comment]
@zoecarver  the changes to  `OSLogOptimization.cpp` and `ConstExpr.h` looks good to me.

**Comment by zoecarver** [comment]
@milseman Will do, thanks for pointing me towards those files.

@gottesmm Sounds good. No rush :)

**Comment by zoecarver** [comment]
I may be running the tests incorrectly but, it said it was going to take about 15 hours (similar estimate after ~3 hours so I killed it). Is there a way to run a single test? Otherwise, can someone trigger the bot?

**Comment by milseman** [comment]
You can invoke lit by hand with a path. If the path is a directory, it will run every test in that directory, and if it's a file just that test. It's a little awkward in that you specify the test path relative to the source directory, but based out of the build directory:

https://github.com/apple/swift/blob/master/docs/Testing.md#using-litpy

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
